File: src/optical_element/MATLAB/Furion_thin_Lens.m
function    out_Field = Furion_thin_Lens(X,Y,in_Field,wavelength,focus_length,flag)
%% ******************************************************************** %%
% By Dr. Chuan Yang,  Email: yangc@mail.iasf.ac.cn
% Institute of Advanced Science Facilities, Shenzhen (IASF)
% July, 2022, Shenzhen
%
% This function is used to calculate the wave going through a thin lens.
% Input:
%    X,Y,in_Field,wavelength,focus_length,flag(=1,=2,3)
%
% output: out_Field
%% ******************************************************************** %%
k = 2*pi/wavelength;
if flag == 1
    out_Field = in_Field.*exp(-1i*k*X.^2/2/focus_length);
elseif flag == 2
    out_Field = in_Field.*exp(-1i*k*Y.^2/2/focus_length);
else
    out_Field = in_Field.*exp(-1i*k*(X.^2+Y.^2)/2/focus_length);
end

File: src/optical_element/Furion_thin_Lens.cpp
#include "Furion_thin_Lens.h"

using namespace Furion_NS;

Furion_Thin_Lens::Furion_Thin_Lens()
{

}

Furion_Thin_Lens::~Furion_Thin_Lens()
{
    
}

void Furion_Thin_Lens::Furion_thin_Lens(std::vector<std::vector<std::complex<double>> >& out_Field, std::vector<std::vector<double> >& X, std::vector<std::vector<double> >& Y, std::vector<std::vector<std::complex<double>> >& in_Field, double wavelength, double focus_length, int flag, int n)
{
    double k = 2 * std::_Pi / wavelength;
    double cache = -k / 2 / focus_length;

    if (flag == 1)
    {
        for (int i = 0; i < n; i++) 
        {
            for (int j = 0; j < n; j++) 
            {
                out_Field[i][j] = in_Field[i][j] * std::exp(std::complex<double>(0, X[i][j] * X[i][j] * cache));
            }
        }
    }
    else if (flag == 2)
    {
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < n; j++)
            {
                out_Field[i][j] = in_Field[i][j] * std::exp(std::complex<double>(0, Y[i][j] * Y[i][j] * cache));
            }
        }
    }
    else
    {
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < n; j++)
            {
                out_Field[i][j] = in_Field[i][j] * std::exp(std::complex<double>(0, Y[i][j] * Y[i][j] * cache));
            }
        }
    }
}




